enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+this [bug](https://issues.redhat.com/browse/OCPBUGS-13219) for details). To
+prevent bugs like this happening in the future and make the solution less
+fragile we should have a test that gates the OpenShift release and that
+verifies that the upgrade can be performed without a registry server. One way
+to ensure this is to run in CI an admission hook that rejects/warns about any spec
+that uses the `Always` pull policy.
+
+It would also be useful to have another test that scans for use of this `Always`
+pull policy in the source code.
+
+We can control the image pull policy for the OCP payload, but not for
+customer-specific images. The OCP upgrade may succeed, but the overall upgrade
+process will still be seen as failed from the customer point of view. For
+example, in the OpenShift appliance use case both the cluster and the customer
+workloads are installed using a temporary registry server that is shutdown
+after the installation is complete. If any of those workloads uses the `Always`
+pull policy then the OCP upgrade would succeed, but the customer workloads will
+not be able to start after the upgrade. To mitagate that risk when a bundle is
+used the upgrade mechanism will check if there are pods using the `Always` pull
+policy, and will emit a warning if there are any.
+
+#### Don't try to contact the image registry server explicitly
+
+Some OpenShift components explicitly try to contact the registry server without
+a fallback alternative. These need to be changed so that they don't do it or so
+that they have a fallback mechanism when the registry server isn't available.
+
+For example, in OpenShift 4.1.13 the machine config operator runs the
+equivalent of `skopeo inspect` in order to decide what kind of upgrade is in
+progress. That fails if there is no registry server, even if the release image
+has already been pulled. That needs to be changed so that contacting the
+registry server is not required. A possible way to do that is to use the
+equivalent of `crictl inspect` instead.
+
+#### MCO should learn to pre-load and pin images
+
+Starting with version 4.14 of OpenShift CRI-O will have the capability to pin
+certain images (see [this](https://github.com/cri-o/cri-o/pull/6862) pull
+request for details). That capability will be used to pin all the images
+required for the upgrade, so that they aren't garbage collected by kubelet and
+CRI-O.
+
+Note that pinning images means that kubelet and CRI-O will not remove them, even
+if they aren't in use. It is very important to make sure that there is enough
+available space for these images, as otherwise the performance of the node may
+degrade and it may stop functioning correctly if it runs out of space. The space
+should be enough to accommodate tho releases (current running + candidate for
+install) as well as workload images and buffer.
+
+The changes to pin the images will be done in a
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, something like this:
```

Comment:
One thing that could actually simplify things here is to make this in `/run/crio/crio.conf` - i.e. it's an explicitly *transient* state before the next boot - and the config automatically goes away after the next boot (when we've upgraded).

I think we should flesh out some more "transient MachineConfig" style flows that write to `/run`, don't reboot, and don't have config drift monitoring etc.